File: cpp/src/bws/packer/InstanceArray.cpp
#include <bws/packer/InstanceArray.hpp>

#include <bws/packer/Wrapper.hpp>

using namespace std;
using namespace bws::packer;
using namespace samchon::library;
using namespace samchon::protocol;

/* ---------------------------------------------------------
	CONSTRUCTORS
--------------------------------------------------------- */
auto InstanceArray::createChild(shared_ptr<XML> xml) -> Instance*
{
	if (xml->getProperty("type") == "product")
		return new Product();
	else
		return new Wrapper();
}

/* ---------------------------------------------------------
	EXPORTERS
--------------------------------------------------------- */
auto InstanceArray::TAG() const -> string
{
	return "instanceArray";
}
auto InstanceArray::CHILD_TAG() const -> string
{
	return "instance";
}

File: ts/src/bws/packer/InstanceArray.ts
/// <reference path="API.ts" />

namespace bws.packer
{
	/**
	 * An array of Instance objects.
	 *
	 * @author Jeongho Nam <http://samchon.org>
	 */
	export class InstanceArray
		extends samchon.protocol.EntityArray<Instance>
	{
		/* -----------------------------------------------------------
			CONSTRUCTORS
		----------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}

		/**
		 * @inheritdoc
		 */
		public createChild(xml: samchon.library.XML): Instance
		{
			let type = xml.getProperty("type");
			
			if (type == "product")
				return new Product();
			else if (type == "wrapper")
				return new Wrapper();
			else
				return null;
		}

		/* -----------------------------------------------------------
			EXPORTERS
		----------------------------------------------------------- */
		/**
		 * @inheritdoc
		 */
		public TAG(): string
		{
			return "instanceArray";
		}

		/**
		 * @inheritdoc
		 */
		public CHILD_TAG(): string
		{
			return "instance";
		}
	}
}